### PR TITLE
GAP 1928 - add logout

### PR DESCRIPTION
**Pull request by john-tco**
Related prs
- https://github.com/cabinetoffice/gap-find-apply-web/pull/49
-https://github.com/cabinetoffice/gap-find-admin-backend/pull/14

user service env pr - https://eu-west-2.console.aws.amazon.com/codesuite/codecommit/repositories/GAP-infra/pull-requests/88/changes?region=eu-west-2

### CODE REVIEW

File: src/main/java/gov/cabinetofice/gapuserservice/util/RestUtils.java [hunk 1]
```diff
@@ -35,6 +35,14 @@ public static JSONObject getRequestWithHeaders(String url, Map<String, String> h
         return convertResponseToJson(response);
     }
 
+    public static JSONObject getRequest(String url) throws IOException {
+        HttpClient httpClient = HttpClients.createDefault();
+        HttpGet httpGet = new HttpGet(url);
+
+        HttpResponse response = httpClient.execute(httpGet);
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetofice/gapuserservice/util/RestUtils.java:42]: We will need some error handling here if One login responds with 401 or 403

File: src/main/java/gov/cabinetofice/gapuserservice/service/OneLoginService.java [hunk 2]
```diff
@@ -282,6 +296,27 @@ public void migrateUser(final User user, final String jwt) {
                 .block();
     }
 
+    public HttpResponse logoutUser(final Cookie customJWTCookie, final HttpServletResponse response) {
+        try {
+            final DecodedJWT decodedJwt = customJwtService.decodedJwt(customJWTCookie.getValue());
+            final JwtPayload payload = customJwtService.decodeTheTokenPayloadInAReadableFormat(decodedJwt);
+
+            HttpResponse res = RestUtils.getRequest(oneLoginLogoutEndpoint + "?id_token_hint=" + payload.getTokenHint() +
```
Comment (ConorFayleTCO) [inline src/main/java/gov/cabinetofice/gapuserservice/service/OneLoginService.java:304]: probably need to sanitise the token here as it's stored in a cookie

File: src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java [hunk 3]
```diff
@@ -106,4 +122,29 @@ public User deleteUser(Integer id) {
...
+            jwtBlacklistService.addJwtToBlacklist(customJWTCookie.getValue());
+        }
+        final Cookie userTokenCookie = WebUtil.buildCookie(
+                new Cookie(userServiceCookieName, null),
+                Boolean.TRUE,
+                Boolean.TRUE,
+                null
+        );
+        userTokenCookie.setMaxAge(0);
+        response.addCookie(userTokenCookie);
```
Comment (ConorFayleTCO) [inline src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java:137]: for this you could use WebUtil.deleteCookie instead

File: src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java [hunk 4]
```diff
@@ -106,4 +122,29 @@ public User deleteUser(Integer id) {
...
+        final String authenticationCookieDomain = Objects.equals(this.configProperties.getProfile(), "LOCAL") ? "localhost" : "cabinetoffice.gov.uk";
+
+        final Cookie thirdPartyAuthToken = WebUtil.buildCookie(
+                new Cookie(authenticationProvider.getTokenCookie(), null),
+                Boolean.TRUE,
+                Boolean.TRUE,
+                authenticationCookieDomain
+        );
+        thirdPartyAuthToken.setMaxAge(0);
+        response.addCookie(thirdPartyAuthToken);
```
Comment (ConorFayleTCO) [inline src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java:148]: could use deleteCookie here as well but might need to overload it to add the domain

File: src/main/java/gov/cabinetofice/gapuserservice/web/LoginControllerV2.java [hunk 5]
```diff
@@ -98,21 +101,22 @@ public RedirectView login(final @RequestParam(name = REDIRECT_URL_NAME) Optional
...
-            final @RequestParam String code,
-            final @RequestParam String state
-    ) {
+             final @RequestParam String code,
+            final @RequestParam String state) {
         final JSONObject tokenResponse = oneLoginService.getOneLoginUserTokenResponse(code);
+        final String authToken = tokenResponse.getString("access_token");
+        final String tokenHint = tokenResponse.getString("id_token");
+
         IdTokenDto decodedIdToken = oneLoginService.getDecodedIdToken(tokenResponse);
```
Comment (ConorFayleTCO) [inline src/main/java/gov/cabinetofice/gapuserservice/web/LoginControllerV2.java:110]: here you could modify `getDecodedIdToken` to instead be `decodeIdToken` which takes the idToken so it's not called `tokenResponse.getString("id_token")` twice - not necessary though

I would rename authToken to accessToken and tokenHint to idToken though for clarity

File: src/test/java/gov/cabinetofice/gapuserservice/web/LoginControllerV2Test.java [hunk 6]
```diff
@@ -385,6 +390,8 @@ void shouldRequireReauthentication_ifNonceIsExpired() throws JSONException {
             final HttpServletResponse response = Mockito.spy(new MockHttpServletResponse());
             final JSONObject tokenResponse = new JSONObject();
             tokenResponse.put("id_token", idToken);
+            tokenResponse.put("tokenHint", "blah");
```
Comment (ConorFayleTCO) [inline src/test/java/gov/cabinetofice/gapuserservice/web/LoginControllerV2Test.java:393]: does this expect a tokenHint? I thought the tokenHint was just the idToken. Same below as well

File: src/test/java/gov/cabinetofice/gapuserservice/web/LoginControllerV2Test.java [hunk 7]
```diff
@@ -385,6 +390,8 @@ void shouldRequireReauthentication_ifNonceIsExpired() throws JSONException {
...
+            tokenResponse.put("access_token", "blah");
```
Comment (ConorFayleTCO) [inline src/test/java/gov/cabinetofice/gapuserservice/web/LoginControllerV2Test.java:394]: access_token value should probably be the same in these 3 tests as they're unrelated to this change

File: src/main/java/gov/cabinetofice/gapuserservice/web/LoginControllerV2.java (hunk 5 above)
Comment (john-tco) [reply src/main/java/gov/cabinetofice/gapuserservice/web/LoginControllerV2.java:110]: Good shout, i'll update that

File: src/test/java/gov/cabinetofice/gapuserservice/web/LoginControllerV2Test.java (hunk 6 above)
Comment (john-tco) [reply src/test/java/gov/cabinetofice/gapuserservice/web/LoginControllerV2Test.java]: Yeah these are not actually needed - have removed

File: src/main/java/gov/cabinetofice/gapuserservice/service/OneLoginService.java [hunk 8]
```diff
@@ -282,14 +297,36 @@ public void migrateUser(final User user, final String jwt) {
...
+        try {
+
+            final DecodedJWT decodedJwt = customJwtService.decodedJwt(customJWTCookie.getValue());
+            final JwtPayload payload = customJwtService.decodeTheTokenPayloadInAReadableFormat(decodedJwt);
+
+            HttpResponse res = RestUtils.getRequest(oneLoginLogoutEndpoint + "?id_token_hint=" + sanitizeCode(payload.getIdToken()) +
+                    "&post_logout_redirect_uri=" + postLogoutRedirectUri);
+
+            if(res.getStatusLine().getStatusCode() == 401 || res.getStatusLine().getStatusCode() == 403) {
+                final Logger logger = null;
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetofice/gapuserservice/service/OneLoginService.java:310]: You don't need to initialise a logger here.

You can use @Log4j2 above the class and then reference log.error(....)

**Review by dylanwrightCO — APPROVED** [review]

**Pull request merged**